src/cgns_internals.c

```diff
@@ -121,6 +124,201 @@ int cgi_read()
+{
+    _childnode_t* p1;
+    _childnode_t* p2;
+
+    p1 = (_childnode_t *) v1;
+    p2 = (_childnode_t *) v2;
+    return p1->type - p2->type;
+}
+
+static int sort_zone_names(const void* v1, const void* v2)
+{
+    _childnode_t* p1;
+    _childnode_t* p2;
+
+    p1 = (_childnode_t*)v1;
+    p2 = (_childnode_t*)v2;
+    return (strcmp(p1->name, p2->name));
+}
+
+#ifdef _WIN32
+#define CG_INLINE __inline
+#else
+#define CG_INLINE static inline
+#endif
+
+CG_INLINE int cgi_get_nodes_with_label(const _childnode_t *childlist, const int nchildren, BaseLabel_t label, int *nnodes, int *startpos)
+{
+    int n;
+    int first = 1;
+    int start = nchildren;
+
+    for (n = 0; n < nchildren; n++) {
+        if (childlist[n].type != label && first) continue;
+        if (childlist[n].type != label) break;
+        if (first) {
+            start = n;
+            first = 0;
+        }
+    }
+    *nnodes = n - start;
+    *startpos = start;
+    return CG_OK;
+}
+
+int cgi_read_all_base_children(double base_id, int* nnodes, _childnode_t** childnodes) {
+    int nid, n, nchildren, len;
+    char nodelabel[CGIO_MAX_NAME_LENGTH + 1];
+    double* idlist;
+    _childnode_t* childlist;
+    *nnodes = 0;
+
+    if (cgio_number_children(cg->cgio, base_id, &nchildren)) {
+        cg_io_error("cgio_number_children");
+        return CG_ERROR;
+    }
+    if (nchildren < 1) return CG_OK;
+    idlist = CGNS_NEW(double, nchildren);
+    if (cgio_children_ids(cg->cgio, base_id, 1, nchildren,
+        &len, idlist)) {
+        cg_io_error("cgio_children_ids");
+        return CG_ERROR;
+    }
+    if (len != nchildren) {
+        CGNS_FREE(idlist);
+        cgi_error("mismatch in number of base children and child IDs read");
+        return CG_ERROR;
+    }
+    childlist = CGNS_NEW(_childnode_t, nchildren);
+    nid = 0;
+    for (nid = 0, n = 0; n < nchildren; n++) {
+        /* Get the node label */
+        if (cgio_get_label(cg->cgio, idlist[n], nodelabel)) {
+            cg_io_error("cgio_get_label");
+            return CG_ERROR;
+        }
+        childlist[n].type = get_base_label_type_as_enum(nodelabel);
+
+        if (childlist[n].type != LabelNull_t) {
+            childlist[nid].id = idlist[n];
+            if (nid < n) {
+                childlist[nid].type = childlist[n].type;
+            }
+            /* Get also the node name */
+            if (cgio_get_name(cg->cgio, idlist[n], childlist[nid].name)) {
+                cg_io_error("cgio_get_name");
+                return CG_ERROR;
+            }
+            nid++;
+        }
+        else {
+            cgio_release_id(cg->cgio, idlist[n]);
+        }
+    }
+    CGNS_FREE(idlist);
+    if (nid > 0) {
+        *childnodes = childlist;
+        *nnodes = nid;
+    }
+    else {
+        CGNS_FREE(childlist);
+    }
+    return CG_OK;
+}
+
+int cgi_read_equations_from_list(int in_link, _childnode_t* nodelist, int nnodes, cgns_equations** equations);
```

Comment:
cgi_read_*_list are private helper function that are only valid for reading CGNSBase_t node
The original function they are derived from are kept since they are possibly called by other part of the code or cgnslib.c